### DISCUSSION TITLE
KubeSpan non-k8s host

### DISCUSSION THREAD

**Discussion by mtricht** [siderolabs/talos#10032] [Q&A]
I am really happy with Talos and KubeSpan so far. However, I have one question that I haven't been able to find an answer to.

I have Node A with a public IP running in the cloud, along with three nodes (B, C, and D) running on bare metal in my homelab behind NAT. I also have Server E, which does not run Talos or Kubernetes, and I would like Node A to be able to communicate with Server E. Is this something that I can achieve using KubeSpan?

**Comment by smira** [comment]
Not at the moment, there's no KubeSpan client except for Talos itself.

You could try running Talos in a container/pod though, and if you expose the Wireguard port outside correctly, and tell Talos about it, it should work.

**Reply by smira** [reply]
Specifically, if you append the following machine configuration document (https://www.talos.dev/v1.9/reference/configuration/network/kubespanendpointsconfig/), Talos in a container would announce that as a KubeSpan Wireguard endpoint. It should point back to the container's KubeSpan port (51820).

**Reply by mtricht** [reply]
Thank you for the pointers. I can get Talos in Docker to join the existing cluster, but I can't get it to work yet. When I apply the machine configuration for additional endpoints, it should show up in the output of `talosctl get kubespanendpoints`, correct?

**Reply by smira** [reply]
No, it shouldn't. You should see that in `talosctl get affiliates -o yaml` for the machine you added the endpoints to.

**Comment by PatrickHuetter** [comment]
I'm interested in this scenario also, since i have this working already using k3s with native tailscale integration. Is it possible to use tailscale instead of kubespan with talos to get a cross datacenter cluster? E.g. installing talos with tailscale and tunneling kubernetes cluster traffic through the private tailscale network?

**Reply by smira** [reply]
Tailscale is available as a system extension, so yes it is possible, but it is certainly more work than KubeSpan to configure things properly.

**Reply by PatrickHuetter** [reply]
@smira Ok, I would even switch to Kubespan if it reduces complexity and works well. The question I’m asking myself is this: I have really different hardware in various locations with different network setups. For example, some servers only have a public IPv4, some have a public dual-stack (IPv4 and IPv6), and some have public IPv6 only. With Tailscale, everything works great so far, because Tailscale handles dual-stack perfectly and works behind firewalls or in NAT environments thanks to NAT traversal. Within the Kubernetes cluster, you don’t even notice any of that, as I have regular dual-stack networking using the cluster and pod CIDRs that just works. Would this be possible with Talos and Kubespan? :)

**Reply by smira** [reply]
I think you should try and tell us, we don't try to compete with Tailscale, so there will be never be a feature parity.

**Reply by PatrickHuetter** [reply]
> I think you should try and tell us, we don't try to compete with Tailscale, so there will be never be a feature parity.

I think i'll try it but currently i'm unsure because of some notes regarding discovery-service in your docs. If i want to run my own discovery service i can't because i must have a commercial license? Is this correct? I understand that i can use the public available discovery from you but what if your server are offline or not reachable some day? Does it affect already running clusters? Is the discovery service needed for kubespan?

**Reply by smira** [reply]
I think these questions are already answered in the docs.

For non-production (homelab), you can run your own discovery service without a license (it's BUSL, see terms).

If the discovery service is down, it doesn't affect KubeSpan immediately (cached data will be used), but it would affect your KubeSpan if e.g. you add or remove machines.
